File: microservices/product/service/src/main/java/fr/edjaz/microservices/core/product/services/MessageProcessor.java
package fr.edjaz.microservices.core.product.services;

import fr.edjaz.api.core.product.Product;
import fr.edjaz.api.core.product.ProductService;
import fr.edjaz.api.event.Event;
import fr.edjaz.util.exceptions.EventProcessingException;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.cloud.stream.annotation.EnableBinding;
import org.springframework.cloud.stream.annotation.StreamListener;
import org.springframework.cloud.stream.messaging.Sink;

@Slf4j
@EnableBinding(Sink.class)
public class MessageProcessor {

    private final ProductService productService;

    @Autowired
    public MessageProcessor(ProductService productService) {
        this.productService = productService;
    }

    @StreamListener(target = Sink.INPUT)
    public void process(Event<Integer, Product> event) {

        LOG.info("Process message created at {}...", event.getEventCreatedAt());

        switch (event.getEventType()) {

        case CREATE:
            Product product = event.getData();
            LOG.info("Create product with ID: {}", product.getProductId());
            productService.createProduct(product);
            break;

        case DELETE:
            int productId = event.getKey();
            LOG.info("Delete recommendations with ProductID: {}", productId);
            productService.deleteProduct(productId);
            break;

        default:
            String errorMessage = "Incorrect event type: " + event.getEventType() + ", expected a CREATE or DELETE event";
            LOG.warn(errorMessage);
            throw new EventProcessingException(errorMessage);
        }

        LOG.info("Message processing done!");
    }
}

File: microservices/product/service/src/main/kotlin/fr/edjaz/microservices/core/product/services/MessageProcessor.kt
package fr.edjaz.microservices.core.product.services

import fr.edjaz.api.core.product.Product
import fr.edjaz.api.core.product.ProductService
import fr.edjaz.api.event.Event
import fr.edjaz.util.exceptions.EventProcessingException
import org.slf4j.LoggerFactory
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import java.util.function.Consumer

@Configuration
class MessageProcessor {
    companion object {
        @Suppress("JAVA_CLASS_ON_COMPANION")
        @JvmStatic
        private val logger = LoggerFactory.getLogger(javaClass.enclosingClass)
    }

    @Bean
    fun sink(productService: ProductService): Consumer<Event<Int, Product>> {
        return Consumer { event: Event<Int, Product> ->
            run {
                logger.info("Process message created at {}...", event.eventCreatedAt)
                when (event.eventType) {
                    Event.Type.CREATE -> {
                        val product = event.data
                        logger.info("Create product with ID: {}", product!!.productId)
                        productService.createProduct(product)
                    }
                    Event.Type.DELETE -> {
                        val productId = event.key!!
                        logger.info("Delete recommendations with ProductID: {}", productId)
                        productService.deleteProduct(productId)
                    }
                    else -> {
                        val errorMessage = "Incorrect event type: " + event.eventType + ", expected a CREATE or DELETE event"
                        logger.warn(errorMessage)
                        throw EventProcessingException(errorMessage)
                    }
                }
                logger.info("Message processing done!")
            }
        }
    }
}
